src/lib/common/payload_uart.cpp

```diff
@@ -28,6 +28,99 @@ static bool _useLineBuffer = true;
 // Flag to use xStreamBufferIsFull or not. If not, user can use LineBuffer.
 static bool _useByteStreamBuffer = false;
 
+// Flag to track if a multi-write transaction is in-progress, useful for eg HALF_DUPLEX
+static bool _transactionInProgress = false;
+// Track if a write occurred during a transaction, to ensure _postTxFunction always called
+static bool _writeDuringTransaction = false;
+// Optional transaction start function
+static HardwareControlFunc _preTxFunction = nullptr;
+// Optional transaction end function, will be called after final write of transaction completes
+static HardwareControlFunc _postTxFunction = nullptr;
+// Semaphore for passing token between postTxCallback, endTransaction, and write to ensure
+//  proper synchronization between completion of all writes in the transaction blocking _postTxFunction.
+static SemaphoreHandle_t _postTxSemaphore = nullptr;
+
+// Forward declaration of the post-transmission callback function
+static void pluartPostTransactionCb(SerialHandle_t *handle);
+
+// Set transaction hardware control functions
+void enableTransactions(HardwareControlFunc preTxFunc, HardwareControlFunc postTxFunc) {
+  // Assign the transaction start and end callbacks
+  _preTxFunction = preTxFunc;
+  _postTxFunction = postTxFunc;
+  // Assign the post-transaction callback function,
+  //  which wraps _postTxFunction in some additional checks to ensure transaction is complete.
+  uart_handle.postTxCb = pluartPostTransactionCb;
+}
+
+// Start a Tx transaction
+// TODO - return false if transaction already started?
+void startTransaction() {
+  taskENTER_CRITICAL();
+  _transactionInProgress = true;
+  _writeDuringTransaction = false;
+  taskEXIT_CRITICAL();
+  if (_preTxFunction) {
+    _preTxFunction(); // External start transaction function. Eg, enable Tx for Half-Duplex
+  }
+}
+
+/** End a Tx transaction, called by user.
+// Blocks until all PLUART writes are complete then fires _postTxFunction
+// Description: First safely check if there's a transaction to end, safely wait for all Tx writing
+                to finish, then _postTxFunction().
+TODO - If users have multiple tasks interacting with PLUART transactions,
+       this will heavily bork. This is not a currently supported use case.
+ **/
+bool endTransaction(uint32_t wait_ms) {
+  taskENTER_CRITICAL();
+  // If not in a transaction, bail and tell the user.
+  if (!_transactionInProgress) {
+    taskEXIT_CRITICAL();
+    return false;
+  }
+
+  // If there were no writes during the transaction, don't wait for write complete semaphore token.
+  if (!_writeDuringTransaction) {
+    _transactionInProgress = false;
+    taskEXIT_CRITICAL();
+    _postTxFunction(); // Safely call the function
```

Comment:
Do we want to check that this is non-NULL before calling?

Reply:
oooo - good catch, thanks @matt001k

Reply:
fix pushed